Use `Tempfile.create`

nobu opened the pull request:
Instead of `Dir::Tmpname.make_tmpname`, an internal method which does not guarantee uniqueness, use `Tempfile.create`.

### Summary

Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together.

### Other Information

If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails!

nobu committed:
Use `Tempfile.create`

Instead of `Dir::Tmpname.make_tmpname`, an internal method which does not guarantee uniqueness, use `Tempfile.create`.

kamipo commented:
Previously `@cache_path` is a directory path generated by `File.join ...`, but `Tempfile.create` will create actual temp file, seems not compatible.

kamipo commented on actionpack/test/controller/log_subscriber_test.rb:
```diff
@@ -98,7 +98,7 @@ def setup
     @old_logger = ActionController::Base.logger
 
-    @cache_path = File.join Dir.tmpdir, Dir::Tmpname.make_tmpname("tmp", "cache")
+    @cache_path = Tempfile.create(%w"tmp cache", Dir.tmpdir)
```
Seems should be here:

```ruby
@cache_path = Tempfile.create(%w"tmp cache", Dir.tmpdir).path
FileUtils.rm(@cache_path)
```

nobu replied:
Thank you for the info.
Then it should use `Dir.mktmpdir`.

nobu committed:
Use `Dir.mktmpdir`

As `@cache_path` is expected to be a directory name, use `Dir.mktmpdir`.
And omit unnecessary `Dir.tmpdir`.

rafaelfranca commented:
Thanks nobu. Merged

yahonda commented:
@rafaelfranca Would you backport this change to `5-1-stable` to address `NoMethodError: undefined method `make_tmpname' for Dir::Tmpname:Module` ? 

https://travis-ci.org/rails/rails/jobs/318406875

yahonda commented:
Thanks for the backport.
